src/v/cluster/topics_frontend.cc

```diff
@@ -949,4 +981,127 @@ topics_frontend::do_cancel_moving_partition_replicas(
+    std::vector<move_topic_replicas_data> cmds;
+
+    auto tp_metadata = _topics.local().get_topic_metadata(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+
+    // units shold exist during replicate_and_wait call
+    absl::btree_map<model::partition_id, allocation_units> units;
+
+    for (const auto& assignment : tp_metadata->get_assignments()) {
+        if (assignment.replicas.size() > new_replication_factor) {
+            co_return errc::topic_invalid_replication_factor;
+        }
+
+        auto partition_constraints = cluster::partition_constraints(
+          assignment.id, new_replication_factor);
+        auto stable_assigments = partition_assignment(
+          assignment.group, assignment.id, assignment.replicas);
+
+        auto ntp = model::ntp(topic.ns, topic.tp, assignment.id);
+
+        auto reallocation = co_await _allocator.invoke_on(
+          partition_allocator::shard,
+          [partition_constraints = std::move(partition_constraints),
+           stable_assigments = std::move(stable_assigments),
+           ntp](partition_allocator& al) {
+              return al.reallocate_partition(
+                partition_constraints,
+                stable_assigments,
+                get_allocation_domain(ntp));
+          });
+
+        if (!reallocation) {
+            vlog(
+              clusterlog.error,
+              "attempt to find reallocation for topic {}, partition {} with "
+              "stable_replicas: {} failed, error: {}",
+              topic,
+              assignment.id,
+              assignment.replicas,
+              reallocation.error().message());
+
+            co_return reallocation.error();
+        }
+
+        units.emplace(assignment.id, std::move(reallocation.value()));
+
+        cmds.emplace_back(
+          assignment.id,
+          units[assignment.id].get_assignments().front().replicas);
+    }
+
+    move_topic_replicas_cmd cmd(topic, std::move(cmds));
+
+    co_return co_await replicate_and_wait(
+      _stm, _features, _as, std::move(cmd), timeout);
+}
+
+ss::future<std::error_code> topics_frontend::decrease_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    std::vector<move_topic_replicas_data> cmds;
+
+    auto tp_metadata = _topics.local().get_topic_metadata_ref(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+
+    for (const auto& assignment : tp_metadata.value().get().get_assignments()) {
+        auto replicas = assignment.replicas;
+        if (replicas.size() < new_replication_factor) {
+            co_return errc::topic_invalid_replication_factor;
+        }
+
+        while (replicas.size() > new_replication_factor) {
+            replicas.pop_back();
+        }
+
+        cmds.emplace_back(assignment.id, std::move(replicas));
```

Comment:
can we use `std::copy_n`, now we are copying the whole replica set which is not necessary